File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      shouting?(input) and question?(input) -> "Calm down, I know what I'm doing!"
      shouting?(input) -> "Whoa, chill out!"
      question?(input) -> "Sure."
      silence?(input) -> "Fine. Be that way!"
      true -> "Whatever."
    end
  end

  defp question?(input), do: String.ends_with?(input, "?")

  defp shouting?(input), do: has_letters?(input) and upper_case?(input)
  defp has_letters?(input), do: String.match?(input, ~r/[\p{L}]/)
  defp upper_case?(input), do: String.upcase(input) == input

  defp silence?(input), do: input |> String.trim() |> Kernel.==("")
end

File: clojure/bob/src/bob.clj
(ns bob
  (:use [clojure.string :only (ends-with? upper-case trim)]))

(def responses {
                #{:question} "Sure."
                #{:shout :question} "Calm down, I know what I'm doing!"
                #{:shout} "Whoa, chill out!"
                #{:silent} "Fine. Be that way!"
                #{} "Whatever." })

(defn- has-letters? [s] (not (empty? (re-find #"[a-zA-Z]" s))))
(defn- is-upper-case? [s] (apply = ((juxt identity upper-case) s)))

(defn- shout [s] (when ((every-pred has-letters? is-upper-case?) s) :shout))
(defn- question [s] (when (ends-with? s "?") :question))
(defn- silent [s] (when (empty? s) :silent))
(def map-labels (juxt question shout silent))
(defn- get-label [s] (->> s map-labels (remove nil?)))

(defn response-for [s]
  (->> s
       trim
       (get-label)
       (reduce conj #{})
       (get responses)))
